File: implementations/fastapi/examples/src/payment/infrastructure/refund_fraud_risk_scorer_native_impl.py
from __future__ import annotations

import random

import numpy as np
from sklearn.linear_model import LogisticRegression

from ..application.service.refund_fraud_risk_scorer import RefundFraudRiskScorer
from ..domain.refund_risk_features import RefundRiskFeatures

TRAINING_EXAMPLE_COUNT = 300
TRAINING_SEED = 42


# Scales each raw feature into a roughly 0-1 range before it reaches the model — the same
# scaling every other language's native RefundFraudRiskScorer implementation (and the shared
# services/fraud-risk-scorer/model.py microservice) applies, so a given feature vector scores
# comparably regardless of which RefundFraudRiskScorer implementation handles it.
def _to_vector(features: RefundRiskFeatures) -> list[float]:
    return [
        features.refund_count_last_30_days / 10,
        features.rejected_refund_count_last_30_days / 5,
        features.refund_to_payment_amount_ratio,
        min(1.0, features.minutes_since_payment / 1440),
    ]


# A synthetic seed dataset standing in for real historical fraud-review outcomes — this
# example has no real user base to draw labeled data from. The label follows an explicit
# ground-truth rule (frequent + high-ratio + fast-after-payment refunds are risky) purely so
# the model has a non-trivial pattern to fit; replace this with real labeled history in
# production. This exact generation rule (TRAINING_EXAMPLE_COUNT=300, TRAINING_SEED=42, the
# same risk_score formula) is mirrored in services/fraud-risk-scorer/model.py and every other
# language's native implementation, so both sides of the "shared service vs. native" pair — and
# every language's own native model — are trained on equivalent data. Deliberately
# self-contained (no import from the repo-root services/ package) so this app has no runtime
# dependency on that directory; only refund_fraud_risk_scorer_http_impl.py calls it, over the
# network.
def _generate_training_data() -> tuple[np.ndarray, np.ndarray]:
    rng = random.Random(TRAINING_SEED)
    vectors: list[list[float]] = []
    labels: list[int] = []

    for _ in range(TRAINING_EXAMPLE_COUNT):
        refund_count = rng.randint(0, 7)
        rejected_count = rng.randint(0, 3)
        ratio = rng.random()
        minutes = rng.random() * 43200
        risk_score = refund_count * 0.15 + rejected_count * 0.3 + ratio * 0.4 + max(0.0, 1 - minutes / 1440) * 0.3
        label = 1 if risk_score > 1.1 else 0
        features = RefundRiskFeatures(
            refund_count_last_30_days=refund_count,
            rejected_refund_count_last_30_days=rejected_count,
            refund_to_payment_amount_ratio=ratio,
            minutes_since_payment=minutes,
        )
        vectors.append(_to_vector(features))
        labels.append(label)

    return np.array(vectors), np.array(labels)


def _train_model() -> LogisticRegression:
    x, y = _generate_training_data()
    model = LogisticRegression()
    model.fit(x, y)
    return model


# Trained once at import time (module import is a one-time event per process) against the
# synthetic dataset above — a module-level global, the same singleton-caching pattern this
# repo's src/database.py (`engine`, `SessionLocal`) already uses in place of a DI container's
# SINGLETON scope (see module-pattern.md). A real deployment would retrain periodically against
# actual refund history instead of training once from a fixed synthetic set at startup.
_model = _train_model()


# A Technical Service (see root docs/architecture/domain-service.md) using scikit-learn
# directly — since this IS a Python implementation, there's no reason to hand-roll gradient
# descent the way a language with no first-class ML library in this repo's dependency set
# would. Self-contained; no extra service needed. See refund_fraud_risk_scorer_http_impl.py for
# the other half of the pair (the shared microservice, called over HTTP).
class RefundFraudRiskScorerNativeImpl(RefundFraudRiskScorer):
    async def score(self, features: RefundRiskFeatures) -> float:
        vector = _to_vector(features)
        return float(_model.predict_proba([vector])[0][1])

File: implementations/java-springboot/examples/src/main/java/com/example/accountservice/payment/infrastructure/RefundFraudRiskScorerNativeImpl.java
package com.example.accountservice.payment.infrastructure;

import com.example.accountservice.payment.application.service.RefundFraudRiskScorer;
import com.example.accountservice.payment.domain.RefundRiskFeatures;
import java.util.ArrayList;
import java.util.List;
import java.util.Random;
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
import org.springframework.stereotype.Component;

/**
 * The default implementation of {@link RefundFraudRiskScorer} — trains a small logistic regression
 * in-process, once, at construction time (this bean is a singleton), against a synthetic dataset
 * that mirrors the ground-truth rule in {@code services/fraud-risk-scorer/model.py} (the shared
 * HTTP alternative, see {@link RefundFraudRiskScorerHttpImpl}), so both sides of the "shared
 * service vs. native" pair are trained on equivalent data. No external ML library — hand-rolled
 * batch gradient descent, deliberately simple/inspectable (see root
 * docs/architecture/domain-service.md's RefundFraudRiskScorer example). Self-contained; no extra
 * service needed, which is why {@code native} is the default ({@code fraud-scorer.mode=native}, see
 * {@code FraudScorerProperties}). A real deployment would retrain periodically against actual
 * refund history instead of training once from a fixed synthetic set at startup.
 */
@Component
@ConditionalOnProperty(
        prefix = "fraud-scorer",
        name = "mode",
        havingValue = "native",
        matchIfMissing = true)
public class RefundFraudRiskScorerNativeImpl implements RefundFraudRiskScorer {

    private static final int TRAINING_EXAMPLE_COUNT = 300;
    private static final long TRAINING_SEED = 42L;
    private static final int FEATURE_COUNT = 4;
    private static final double LEARNING_RATE = 0.5;
    private static final int EPOCHS = 500;

    private record TrainingExample(RefundRiskFeatures features, double label) {}

    private record Model(double[] weights, double bias) {}

    // Trained once, here, at construction — reused for every score() call for the lifetime of
    // this singleton bean.
    private final Model model = trainLogisticRegression(generateTrainingData());

    @Override
    public double score(RefundRiskFeatures features) {
        double[] vector = toVector(features);
        double z = model.bias();
        for (int i = 0; i < vector.length; i++) {
            z += vector[i] * model.weights()[i];
        }
        return sigmoid(z);
    }

    // A synthetic seed dataset standing in for real historical fraud-review outcomes — this
    // example has no real user base to draw labeled data from. The label follows an explicit
    // ground-truth rule (frequent + high-ratio + fast-after-payment refunds are risky) purely so
    // the model has a non-trivial pattern to fit; replace this with real labeled history in
    // production. java.util.Random(42) makes the generated dataset — and therefore the trained
    // weights — identical on every run.
    private static List<TrainingExample> generateTrainingData() {
        Random random = new Random(TRAINING_SEED);
        List<TrainingExample> examples = new ArrayList<>(TRAINING_EXAMPLE_COUNT);
        for (int i = 0; i < TRAINING_EXAMPLE_COUNT; i++) {
            int refundCount = random.nextInt(8);
            int rejectedCount = random.nextInt(4);
            double ratio = random.nextDouble();
            double minutes = random.nextDouble() * 43200;
            double riskScore =
                    refundCount * 0.15
                            + rejectedCount * 0.3
                            + ratio * 0.4
                            + Math.max(0, 1 - minutes / 1440) * 0.3;
            double label = riskScore > 1.1 ? 1 : 0;
            examples.add(
                    new TrainingExample(
                            new RefundRiskFeatures(refundCount, rejectedCount, ratio, minutes),
                            label));
        }
        return examples;
    }

    // Scales each raw feature into a roughly 0-1 range before it reaches the model — plain
    // gradient descent converges far more slowly (and less reliably) on unscaled inputs this
    // different in magnitude (a single-digit count next to a value in the thousands). Mirrored
    // exactly in services/fraud-risk-scorer/model.py's to_vector.
    private static double[] toVector(RefundRiskFeatures features) {
        return new double[] {
            features.refundCountLast30Days() / 10.0,
            features.rejectedRefundCountLast30Days() / 5.0,
            features.refundToPaymentAmountRatio(),
            Math.min(1, features.minutesSincePayment() / 1440)
        };
    }

    private static double sigmoid(double z) {
        return 1 / (1 + Math.exp(-z));
    }

    // Batch gradient descent on plain logistic regression — no external ML library, deliberately
    // simple/inspectable. This is the "each language trains natively" side of the pair;
    // RefundFraudRiskScorerHttpImpl (the shared services/fraud-risk-scorer microservice) is the
    // "one shared service" side.
    private static Model trainLogisticRegression(List<TrainingExample> examples) {
        double[] weights = new double[FEATURE_COUNT];
        double bias = 0;
        int n = examples.size();

        for (int epoch = 0; epoch < EPOCHS; epoch++) {
            double[] weightGradients = new double[FEATURE_COUNT];
            double biasGradient = 0;

            for (TrainingExample example : examples) {
                double[] vector = toVector(example.features());
                double z = bias;
                for (int i = 0; i < vector.length; i++) {
                    z += vector[i] * weights[i];
                }
                double prediction = sigmoid(z);
                double error = prediction - example.label();
                for (int i = 0; i < vector.length; i++) {
                    weightGradients[i] += error * vector[i];
                }
                biasGradient += error;
            }

            for (int i = 0; i < weights.length; i++) {
                weights[i] -= (LEARNING_RATE * weightGradients[i]) / n;
            }
            bias -= (LEARNING_RATE * biasGradient) / n;
        }

        return new Model(weights, bias);
    }
}
